### PR TITLE
Add Questionnaire ID

### PR DESCRIPTION
**Pull request by iwootten**
### What is the context of this PR?

Introduce the questionnaire id as part of the required claims in the metadata passed to EQ and pass it on to submission payloads. If case id is present in the list of claims, pass this on too.

https://trello.com/c/0fkw3X5F/2984-jwt-claims-questionnaireid-and-caseid

### How to review 

You need to use the following launcher PR to launch a survey: https://github.com/ONSdigital/eq-questionnaire-launcher/pull/2

### CODE REVIEW

File: app/storage/metadata_parser.py [hunk 1]
```diff
@@ -47,22 +52,24 @@ def clean_leading_trailing_spaces(metadata):
     'boolean': boolean_parser,
     'string': string_parser,
     'optional_string': optional_string_parser,
+    'optional_uuid': optional_uuid_4_parser,
 }
 
-MANDATORY_METADATA = [
+METADATA_VALIDATION_RULES = [
```
Comment (MebinAbraham) [inline app/storage/metadata_parser.py]: We should really rebase against Rich's changes as this is overlapping and has changed.

File: app/views/feedback.py [hunk 2]
```diff
@@ -71,8 +71,9 @@ def send_feedback():
         )
         sent = current_app.eq['submitter'].send_message(
             encrypted_message,
-            case_id=metadata.get('case_id'),
             tx_id=metadata.get('tx_id'),
+            questionnaire_id=metadata.get('questionnaire_id'),
+            case_id=metadata.get('case_id', None),
```
Comment (MebinAbraham) [inline app/views/feedback.py]: Redundant `None`.

File: app/views/flush.py [hunk 3]
```diff
@@ -66,8 +66,9 @@ def _submit_data(user):
 
         sent = current_app.eq['submitter'].send_message(
             encrypted_message,
-            case_id=metadata.get('case_id'),
             tx_id=metadata.get('tx_id'),
+            questionnaire_id=metadata.get('questionnaire_id'),
+            case_id=metadata.get('case_id', None),
```
Comment (MebinAbraham) [inline app/views/flush.py]: Same here, don't need to specify a default.

File: app/views/questionnaire.py [hunk 4]
```diff
@@ -550,7 +551,10 @@ def submit_answers(routing_path, schema):
         message, current_app.eq['key_store'], KEY_PURPOSE_SUBMISSION
     )
     sent = current_app.eq['submitter'].send_message(
-        encrypted_message, case_id=metadata.get('case_id'), tx_id=metadata.get('tx_id')
+        encrypted_message,
+        questionnaire_id=metadata.get('questionnaire_id'),
+        case_id=metadata.get('case_id', None),
```
Comment (MebinAbraham) [inline app/views/questionnaire.py]: And here

File: app/storage/metadata_parser.py (hunk 1 above)
Comment (iwootten) [reply app/storage/metadata_parser.py]: Those changes aren't yet approved right?

Comment (MebinAbraham) [reply app/storage/metadata_parser.py]: Nope, I meant more when it is in?.

Comment (iwootten) [reply app/storage/metadata_parser.py]: Sure, but I can't see how that's different to our usual workflow.

Comment (MebinAbraham) [reply app/storage/metadata_parser.py]: I mean from the point that this card has overlaps with Rich's and maybe should have been on hold, as reviewing metadata parser is essentially redundant if/when Rich's goes in?.

Comment (iwootten) [reply app/storage/metadata_parser.py]: The card was originally on hold, and subsequently marked as unblocked - I consulted with Rich on whether he felt it ready to start. Yes they overlap, but then as I understand it, this card was originally marked the higher priority.

File: app/storage/metadata_parser.py [hunk 5]
```diff
@@ -55,10 +55,11 @@ class RunnerMetadataSchema(Schema, StripWhitespaceMixin):
     ru_ref = VALIDATORS['string'](validate=validate.Length(min=1))
     collection_exercise_sid = VALIDATORS['string'](validate=validate.Length(min=1))
     tx_id = VALIDATORS['uuid']()
-    case_id = VALIDATORS['uuid']()
+    questionnaire_id = VALIDATORS['string']()
```
Comment (MebinAbraham) [inline app/storage/metadata_parser.py]: We should add length to these as currently, empty strings would be valid.

**Review by MebinAbraham — CHANGES_REQUESTED** [review]

File: app/storage/metadata_parser.py (hunk 5 above)
Comment (iwootten) [reply app/storage/metadata_parser.py]: I've changed these for now, but for case_id/questionnaire_id - I wonder if we should be validating them at all. Given that we don't use them for anything - we probably should just be passing them on?

Comment (MebinAbraham) [reply app/storage/metadata_parser.py]: Possibly, guess it depends on how mandatory it actually is. Do we return a 403 if they don't provide us a questionnaire_id vs rm to preventing launching without questionnaire_id.

Comment (iwootten) [reply app/storage/metadata_parser.py]: questionnaire_id will cause a failure, case_id will not

Comment (MebinAbraham) [reply app/storage/metadata_parser.py]: Ya I think your change is fine to validate questionnaire_id/response_id since they are needed but not case_id

**Review by ajmaddaford — APPROVED** [review]

**Review by bitdivision — COMMENTED** [review]
Looks great :+1:

**Review by bitdivision — APPROVED** [review]

**Review by MebinAbraham — APPROVED** [review]

**Review by jonnyshaw89 — APPROVED** [review]

**Pull request merged**